vshard/storage/init.lua

```diff
@@ -361,12 +366,63 @@ local function bucket_commit_delete(bucket)
 end
 
 --
--- Validate whether _bucket update is possible. Bad ones need to be rejected for
--- the sake of keeping data consistency. Ideally it should never throw. But
--- there sure are bugs and this trigger should help to catch them from time to
--- time.
+-- Ensure that a bucket status update is one of these:
+--  * none -> RECEIVING
+--  * RECEIVING -> GARBAGE
+--  * RECEIVING -> ACTIVE
+--  * PINNED <-> ACTIVE <-> SENDING -> SENT -> GARBAGE -> none
 --
-local function bucket_prepare_update(old_bucket, new_bucket)
+local function bucket_validate_update_transition(old_bucket, new_bucket)
+    local new_status = new_bucket.status
+    local bid = new_bucket.id
+    if old_bucket == nil then
+        if new_status ~= BSTATE.RECEIVING then
+            bucket_reject_update(bid, "can't create bucket with '%s' status",
+                                 new_status)
+        end
+        return
+    end
+    local old_status = old_bucket.status
+    if new_status == old_status then
+        return
+    end
+    if new_status == BSTATE.RECEIVING then
+        bucket_reject_update_transition(bid, old_status, new_status)
+    end
+    if new_status == BSTATE.GARBAGE then
+        if old_status ~= BSTATE.RECEIVING and old_status ~= BSTATE.SENT then
+            bucket_reject_update_transition(bid, old_status, new_status)
+        end
+        return
+    end
+    if new_status == BSTATE.PINNED or new_status == BSTATE.SENDING then
+        if old_status ~= BSTATE.ACTIVE then
+            bucket_reject_update_transition(bid, old_status, new_status)
+        end
+        return
+    end
+    if new_status == BSTATE.ACTIVE then
+        if old_status ~= BSTATE.RECEIVING and old_status ~= BSTATE.PINNED and
+                old_status ~= BSTATE.SENDING then
+            bucket_reject_update_transition(bid, old_status, new_status)
+        end
+        return
+    end
+    if new_status == BSTATE.SENT then
+        if old_status ~= BSTATE.SENDING then
+            bucket_reject_update_transition(bid, old_status, new_status)
+        end
+        return
+    end
+    bucket_reject_update(bid, "unknown bucket status: '%s'", new_status)
+end
```

Comment:
Ah, I remember I tried to do the same. But then it appeared that much easier it is to build a state machine:
<details><summary>Diff</summary>

```Diff
diff --git a/vshard/storage/init.lua b/vshard/storage/init.lua
index 892c9cd..15df4bf 100644
--- a/vshard/storage/init.lua
+++ b/vshard/storage/init.lua
@@ -226,6 +226,26 @@ else
     end
 end
 
+local bucket_state_edges = {
+    -- From nothing. Have to cast nil to box.NULL explicitly then.
+    [box.NULL] = {BSTATE.RECEIVING},
+    [BSTATE.ACTIVE] = {BSTATE.PINNED, BSTATE.SENDING},
+    [BSTATE.PINNED] = {BSTATE.ACTIVE},
+    [BSTATE.SENDING] = {BSTATE.ACTIVE, BSTATE.SENT},
+    [BSTATE.SENT] = {BSTATE.GARBAGE},
+    [BSTATE.RECEIVING] = {BSTATE.GARBAGE, BSTATE.ACTIVE},
+    [BSTATE.GARBAGE] = {box.NULL},
+}
+-- Turn the edges into a table where edges[src][dst] would be true if the
+-- src->dst transition is allowed.
+for src, dsts in pairs(bucket_state_edges) do
+    local dst_dict = {}
+    for _, dst in pairs(dsts) do
+        dst_dict[dst] = true
+    end
+    bucket_state_edges[src] = dst_dict
+end
+
 --
 -- Invoke a function on this instance. Arguments are unpacked into the function
 -- as arguments.
@@ -321,11 +341,6 @@ local function bucket_reject_update(bid, message, ...)
                         message:format(...)))
 end
 
-local function bucket_reject_update_transition(bid, old_status, new_status)
-    bucket_reject_update(bid, "bucket transition from '%s' to '%s' " ..
-                         "is prohibited", old_status, new_status)
-end
-
 --
 -- Make _bucket-related data up to date with the committed changes.
 --
@@ -366,72 +381,28 @@ local function bucket_commit_delete(bucket)
 end
 
 --
--- Ensure that a bucket status update is one of these:
---  * none -> RECEIVING
---  * RECEIVING -> GARBAGE
---  * RECEIVING -> ACTIVE
---  * PINNED <-> ACTIVE <-> SENDING -> SENT -> GARBAGE -> none
+-- Validate whether _bucket update is possible. Bad ones need to be rejected for
+-- the sake of keeping data consistency. Ideally it should never throw. But
+-- there sure are bugs and this trigger should help to catch them from time to
+-- time.
 --
-local function bucket_validate_update_transition(old_bucket, new_bucket)
+local function bucket_prepare_update(old_bucket, new_bucket)
     local new_status = new_bucket.status
+    local old_status = old_bucket ~= nil and old_bucket.status or nil
     local bid = new_bucket.id
-    if old_bucket == nil then
-        if new_status ~= BSTATE.RECEIVING then
-            bucket_reject_update(bid, "can't create bucket with '%s' status",
-                                 new_status)
+    if new_status ~= old_status then
+        local src = bucket_state_edges[old_status]
+        if not src then
+            bucket_reject_update(bid, "unknown bucket state '%s'", old_status)
+        elseif not src[new_status] then
+            bucket_reject_update(bid, "transition '%s' to '%s' is not allowed",
+                                 old_status, new_status)
         end
-        return
-    end
-    local old_status = old_bucket.status
-    if new_status == old_status then
-        return
-    end
-    if new_status == BSTATE.RECEIVING then
-        bucket_reject_update_transition(bid, old_status, new_status)
-    end
-    if new_status == BSTATE.GARBAGE then
-        if old_status ~= BSTATE.RECEIVING and old_status ~= BSTATE.SENT then
-            bucket_reject_update_transition(bid, old_status, new_status)
-        end
-        return
-    end
-    if new_status == BSTATE.PINNED or new_status == BSTATE.SENDING then
-        if old_status ~= BSTATE.ACTIVE then
-            bucket_reject_update_transition(bid, old_status, new_status)
-        end
-        return
-    end
-    if new_status == BSTATE.ACTIVE then
-        if old_status ~= BSTATE.RECEIVING and old_status ~= BSTATE.PINNED and
-                old_status ~= BSTATE.SENDING then
-            bucket_reject_update_transition(bid, old_status, new_status)
-        end
-        return
     end
-    if new_status == BSTATE.SENT then
-        if old_status ~= BSTATE.SENDING then
-            bucket_reject_update_transition(bid, old_status, new_status)
-        end
-        return
-    end
-    bucket_reject_update(bid, "unknown bucket status: '%s'", new_status)
-end
-
-
---
--- Validate whether _bucket update is possible according to the already
--- existing refs. Prevent data invalidation if request is in progress.
---
-local function bucket_validate_update_ref(old_bucket, new_bucket)
-    local new_status = new_bucket.status
-    local bid = new_bucket.id
     local ref = M.bucket_refs[bid]
     if old_bucket == nil and ref ~= nil then
         bucket_reject_update(bid, "new bucket can't have a ref object")
     end
-    if new_status == BSTATE.PINNED or new_status == BSTATE.ACTIVE then
-        return
-    end
     if new_status == BSTATE.SENDING or new_status == BSTATE.SENT then
         if ref ~= nil and ref.rw > 0 then
             bucket_reject_update(bid, "'%s' bucket can't have rw refs",
@@ -446,22 +417,18 @@ local function bucket_validate_update_ref(old_bucket, new_bucket)
         end
         return
     end
-end
-
---
--- Validate whether _bucket update is possible. Bad ones need to be rejected for
--- the sake of keeping data consistency. Ideally it should never throw. But
--- there sure are bugs and this trigger should help to catch them from time to
--- time.
---
-local function bucket_prepare_update(old_bucket, new_bucket)
-    bucket_validate_update_ref(old_bucket, new_bucket)
-    bucket_validate_update_transition(old_bucket, new_bucket)
+    bucket_reject_update(bid, "unknown bucket status: '%s'", new_status)
 end
 
 local function bucket_prepare_delete(bucket)
     local bid = bucket.id
     local status = bucket.status
+    local src = bucket_state_edges[status]
+    if not src then
+        bucket_reject_update(bid, "unknown bucket state '%s'", status)
+    elseif not src[box.NULL] then
+        bucket_reject_update(bid, "can't delete '%s' bucket", status)
+    end
     local ref = M.bucket_refs[bid]
     if ref ~= nil and (ref.ro > 0 or ref.rw > 0) then
         bucket_reject_update(bid, "can't delete a bucket with refs")
```
</details>

Reply:
Such a great solution. Hope, you wouldn't mind if I borrow it)